3.next - Make entity context recognize junction tables

ndm2 opened the pull request:
Adds support to the entity context for recognizing `_joinData` properties and resolving the respective junction table classes. This makes reading errors, introspecting types, etc working with join table columns.

While testing I noticed that the root fallback behavior in [**`EntityContext::_getTable()`**](https://github.com/cakephp/cakephp/blob/9fff233e7d38e0f30dd55377af637775f7f98104/src/View/Form/EntityContext.php#L483) doesn't actually falls back to the root node, but the last found / parent node.

The newly added `testTypeAssociatedJoinData` test seems to be the only one that fails when the fallback behavior is being changed to actually fall back to the root node, ie it seems that there is no other test actually covering that functionality, so I'm not 100% sure whether the current behavior is the actually intended behavior, and the description/name should be changed, or if it's maybe a bug, and it should really fall back to the root node?

Commit:
Make entity context recognize junction tables.

Adds support to the entity context for recognizing `_joinData`
properties and resolving the respective junction table classes. This
makes reading errors, introspecting types, etc working with join table
columns.

Commit:
Fix primary key test.

Without tables/schemas set up properly, the primary key check will
always use the `Articles` tables schema for retrieving the foreign
key constraint.

Commit:
Ensure tests use a cross DBMS compatible schema.

markstory reviewed (approved):
Looks good to me. Nice work on the tests.

markstory commented:
>  I'm not 100% sure whether the current behavior is the actually intended behavior, and the description/name should be changed, or if it's maybe a bug, and it should really fall back to the root node?

When properties are undefined I think the correct behavior is to fallback to the farthest table that could be reached in the path. The 'root' name is misleading and wasn't changed when support for working with array fields was added/fixed.

ndm2 commented:
I see... I'll squeeze a change for that in this PR, argument name changes should be backwards compatible IIRC.

Commit:
Cleanup

Remove copy pasta, make use of existing variable, update fallback
argument name and description, and additional assertions to make sure
the non-`_joinData` fallback case is properly covered.

lorenzo commented:
Great job @ndm2 !
